pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,220 @@
+    except VirtualFolderTreeItem.DoesNotExist:
+        parts = split_pootle_path(store.parent.pootle_path)
+        path_parts = ["", parts[0], parts[1], vf.name]
+        if parts[2]:
+            path_parts.append(parts[2].strip("/"))
+        path_parts.append(parts[3])
+        pootle_path = "/".join(path_parts)
+        vfolder_treeitem = (
+            VirtualFolderTreeItem.objects.create(
+                pootle_path=pootle_path,
+                directory=store.parent,
+                vfolder=vf))
+        created = True
+    if not created:
+        # The VirtualFolderTreeItem already existed, so
+        # calculate again the stats up to the root.
+        vfolder_treeitem.update_all_cache()
+
+
+class VirtualFolderFinder(object):
+    """Find vfs for a new store"""
+
+    def __init__(self, store):
+        self.store = store
+
+    @property
+    def language(self):
+        return self.store.translation_project.language
+
+    @property
+    def project(self):
+        return self.store.translation_project.project
+
+    @property
+    def possible_vfolders(self):
+        return (
+            self.project.vfolders.filter(all_languages=True)
+            | self.language.vfolders.filter(all_projects=True)
+            | self.project.vfolders.filter(languages=self.language)
+            | VirtualFolder.objects.filter(
+                all_languages=True, all_projects=True))
+
+    def add_to_vfolders(self):
+        to_add = []
+        for vf in self.possible_vfolders:
+            if vf.path_matcher.should_add_store(self.store):
+                to_add.append(vf)
+        if to_add:
+            self.store.vfolders.add(*to_add)
+            self.store.set_priority()
+        for vf in to_add:
+            update_vfolder_tree(vf, self.store)
+
+
+class VirtualFolderPathMatcher(object):
+
+    tp_path = "/[^/]*/[^/]*/"
+
+    def __init__(self, vf):
+        self.vf = vf
+
+    @property
+    def existing_stores(self):
+        """Currently associated Stores"""
+        return self.vf.stores.all()
+
+    @property
+    def languages(self):
+        """The languages associated with this vfolder
+        If `all_languages` is set then `None` is returned
+        """
+        if self.vf.all_languages:
+            return None
+        return self.vf.languages.values_list("pk", flat=True)
+
+    @property
+    def projects(self):
+        """The projects associated with this vfolder
+        If `all_projects` is set then `None` is returned
+        """
+        if self.vf.all_projects:
+            return None
+        return self.vf.projects.values_list("pk", flat=True)
+
+    @property
+    def matching_stores(self):
+        """Store qs containing all stores that match
+        project, language, and rules for this vfolder
+        """
+        return self.filter_from_rules(self.store_qs)
+
+    @property
+    def rules(self):
+        """Glob matching rules"""
+        return (
+            "%s" % r.strip()
```

Comment:
im going to move these to json and we can add validation then - for now i think a comma-separated string that is not ws-tolerant in this way is a bad idea